### DISCUSSION TITLE
Best way to turn JSON into page content.

### DISCUSSION THREAD

**Discussion by Andre601** [mkdocs/mkdocs#3487] [Q&A] [answered]
I would like to turn the content of a JSON file into some page content, preferably through a hook.

The reason I want to make it through a hook is, that it would allow me to use a placeholder to replace, therefore having the freedom of choosing where it would be, rather than having rigit template placements (if jinja2 placement was another option here).

I would also be open for using javascript here.

The JSON itself has the structure similar to this:
```json
{
  "prompts": [
    "string",
    "another String",
    "Yet another String"
  ]
}
```

**Comment by Andre601** [comment]
Also, for a bit more clarity, I would like to change this list of entries into a single-column table (Only using `<tbody>` without a `<thead>`)

**Comment by oprypin** [comment]
This is well suited for Jinja templates. You can put those into Markdown pages by using the Macros plugin, or enable them only for 1 page manually as done here https://github.com/mkdocs/mkdocs/commit/c6756222808d48fb587d321807674d27dab3f801#diff-4d108d1a11907ec35a9358394f8cc0e3422e52e37b8264922862a8aa4edf4d66

**Reply by Andre601** [reply]
And would that then just be `{% for prompt in prompts %}` to access the JSON or what else would I need to make sure of to get the file content?

Also, I may consider using some frontmatter values here to have it more dynamic (i.e. set where to get the file from and to enable this stuff...), so what would be a setup for that in the hooks?

**Reply by Andre601** [reply]
I tried a setup using parts from squidfunk's `shortcode.py` hook to resolve paths, but I get a `FileNotFoundError: [Errno 2] No such file or directory: '../../assets/extra_quips.json` error.

The JSON file is located inside the `assets` folder (in docs folder) and the path is relative to the file it was used in and does resolve to the file path... Not sure why it treats it as the file not being found...

The current code I have:
```py
import jinja2
import posixpath
import json

from mkdocs.structure.files import File, Files
from mkdocs.structure.pages import Page
from mkdocs.plugins import get_plugin_logger

log = get_plugin_logger("Quip Prompts Renderer")

def on_page_markdown(markdown: str, page: Page, files: Files, **kwargs):
    if page.file.src_uri != 'twitch-minigames/common/quip-battle.md':
        return
    
    path = _resolve_path('assets/extra_quips.json', page, files)
    log.info(path)
    with open(path) as json_file:
        json_content = json.loads(json_file)
    
    return jinja2.Template(markdown).render(json = json_content)

def _resolve_path(path: str, page: Page, files: Files):
    path = _resolve(path, page)
    return path

def _resolve(file_path: str, page: Page):
    path = posixpath.relpath(file_path, page.file.src_uri)
    return posixpath.sep.join(path.split(posixpath.sep)[1:])
```

**Reply by kamilkrzyskow** [reply, marked as answer]
You're resolving the relative path to the `quip-battle.md` page, and omit some first part in `path.split(posixpath.sep)[1:]` 🤔, however the Python command `mkdocs serve` does run from the root directory. This means that the resolved relative path is used relative from the root directory and therefore it doesn't work.

As I understand it, the `assets/extra_quips.json` file is really `docs/assets/extra_quips.json`, right? 
Instead of resolving the path per page, just access the `config.docs_dir` value, as the `on_page_markdown` does have access to it in the kwargs https://www.mkdocs.org/dev-guide/plugins/#on_page_markdown

Also I recommend to explicitly state the encoding in the open function:
```py
path = posixpath.sep.join([config.docs_dir, "assets/extra_quips.json"])
with open(path, encoding="utf8") as json_file:`
```
Note that I never used posixpath before, just followed your example above, with thet `posixpath.sep.join`, there might be a better way 🤷

**Reply by Andre601** [reply]
That worked. Thanks.
